core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,257 @@
+	/// up an initial runtime instance. The parameter is only needed for calling
+	/// into the Wasm module to find out the `Core_version`.
+	///
+	/// `default_heap_pages` - Default number of 64KB pages to allocate for
+	/// Wasm execution. Defaults to `DEFAULT_HEAP_PAGES` if `None` is provided.
+	///
+	/// `maybe_requested_version` - If `Some(RuntimeVersion)` is provided the
+	/// cached instance will be checked for compatibility. In case of incompatibility
+	/// the instance will be reset and a new one will be created synchronously.
+	///
+	/// # Return value
+	///
+	/// If no error occurred a `RuntimePreproc::ValidCode` is returned, containing
+	/// a wasmi `ModuleRef` with an optional `RuntimeVersion` (if the call
+	/// `Core_version` returned a version).
+	///
+	/// In case an error occurred `RuntimePreproc::InvalidCode` is returned with an
+	/// appropriate description.
+	pub fn fetch_runtime<E: Externalities<Blake2Hasher>>(
+		&mut self,
+		wasm_executor: &WasmExecutor,
+		ext: &mut E,
+		default_heap_pages: Option<u64>,
+		maybe_requested_version: Option<&RuntimeVersion>,
+	) -> Result<(WasmModuleInstanceRef, Option<RuntimeVersion>)> {
+		if let None = self.runtime_instance {
+			trace!(target: "runtimes_cache",
+				   "no instance found in cache, creating now.");
+			self.create_instance(wasm_executor, ext, default_heap_pages);
+		}
+
+		let action = match maybe_requested_version {
+			None => Action::ReuseInstance,
+			Some(ref requested_version) => {
+				match self.runtime_instance {
+					Some(RuntimePreproc::InvalidCode) => Action::InvalidCode,
+					Some(RuntimePreproc::ValidCode(_, None)) => {
+						// In case of a certain version being requested, but
+						// the runtime in the cache not having one, we create
+						// a new instance.
+						Action::CreateNewInstance
+					},
+					Some(RuntimePreproc::ValidCode(_, Some(ref runtime_version))) => {
+						if runtime_version.can_call_with(&requested_version) {
+							Action::ReuseInstance
+						} else {
+							trace!(target: "runtimes_cache",
+								   "resetting cache because new version received");
+							Action::CreateNewInstance
+						}
+					},
+					None => unreachable!("instance is created at beginning of function if non-existent; qed"),
+				}
+			},
+		};
+
+		let runtime_preproc = match action {
+			Action::ReuseInstance => {
+				self.reset_instance();
+				self.runtime_instance.clone()
+					.expect("this path will only be invoked if instance exists; qed")
+			},
+			Action::CreateNewInstance => {
+				self.create_instance(wasm_executor, ext, default_heap_pages);
+				self.runtime_instance.clone()
+					.expect("was created right beforehand; qed")
+			},
+			Action::InvalidCode => {
+				RuntimePreproc::InvalidCode
+			},
+		};
+
+		match runtime_preproc {
+			RuntimePreproc::InvalidCode => {
+				let code = ext.original_storage(well_known_keys::CODE).unwrap_or(vec![]);
+				Err(Error::InvalidCode(code))
```

Comment:
This error enum is defined `[here](https://github.com/paritytech/substrate/blob/master/core/executor/src/error.rs#L40-L42)` and requires the code, I just extracted this part from `native_executor`. Not sure about the reasoning, but since the original code later outputs the `code` in a `trace!` this could have been a reason.

Reply:
I am not sure about usability of this and I am going to remove this.